bpo-31432: Clarify CERT_NONE/OPTIONAL/REQUIRED doc

tiran opened the pull request:
The documentation for CERT_NONE, CERT_OPTIONAL, and CERT_REQUIRED were
misleading and partly wrong. It fails to explain that OpenSSL behaves
differently in client and server mode. Also OpenSSL does validate the
cert chain everytime. With SSL_VERIFY_NONE a validation error is not
fatal in client mode and does not request a client cert in server mode.

Also discourage people from using CERT_OPTIONAL in client mode.

Signed-off-by: Christian Heimes <christian@python.org>

<!-- issue-number: bpo-31432 -->
https://bugs.python.org/issue31432
<!-- /issue-number -->

tiran committed:
bpo-31432: Clarify CERT_NONE/OPTIONAL/REQUIRED doc

The documentation for CERT_NONE, CERT_OPTIONAL, and CERT_REQUIRED were
misleading and partly wrong. It fails to explain that OpenSSL behaves
differently in client and server mode. Also OpenSSL does validate the
cert chain everytime. With SSL_VERIFY_NONE a validation error is not
fatal in client mode and does not request a client cert in server mode.

Also discourage people from using CERT_OPTIONAL in client mode.

Signed-off-by: Christian Heimes <christian@python.org>

tiran commented:
Memo to me: drop ``PROTOCOL_TLS_CLIENT`` in 2.7 backport.

alex commented on Modules/_ssl.c:
```diff
@@ -3228,8 +3228,8 @@ set_check_hostname(PySSLContext *self, PyObject *arg, void *c)
         PyErr_SetString(PyExc_ValueError,
-                        "check_hostname needs a SSL context with either "
-                        "CERT_OPTIONAL or CERT_REQUIRED");
+                        "check_hostname needs a SSLContext with "
```
nit: should be `an SSLContext`

vadmium commented on Doc/library/ssl.rst:
```diff
@@ -563,20 +563,28 @@ Constants
-   are provided, validation will be attempted and an :class:`SSLError`
-   will be raised on failure.
+   parameter to :func:`wrap_socket`.  In client mode, :const:`CERT_OPTIONAL`
+   has the same meaning as :const:`CERT_REQUIRED`. It is recommended to
```
What if “anonymous ciphers are enabled”, mentioned under <https://docs.python.org/dev/library/ssl.html#verifying-certificates>?

tiran replied:
I'd rather not explain anonymous cipher suites and just remove them from the documentation completely. Anonymous cipher suites is a misleading term. They don't provide anonymity for the user but allow the server to stay anonymous by not providing any authentication. It's totally insecure and no longer supported in TLS 1.3 for a very good reason.

vadmium commented on Doc/library/ssl.rst:
```diff
@@ -563,20 +563,28 @@ Constants
-   If a certificate is received from the other end, no attempt to validate it
-   is made.
+   parameter to :func:`wrap_socket`.  Except for :const:`PROTOCOL_TLS_CLIENT`,
+   it is the default mode.  With client-side sockets, just about any
```
What does _cert_reqs_ default to if I call wrap_socket(..., ssl_version=PROTOCOL_TLS_CLIENT)? Judging by the signature and documentation for _wrap_socket_, CERT_NONE seems to be the default regardless of the protocol setting. If you are only talking about the default value for _verify_mode_ after an SSLContext is constructed, I suggest to mention that under the entry for SSLContext.verify_mode, or the SSLContext constructor, instead of here. (Otherwise, maybe adjust the description of the top-level _wrap_socket_ function.)

tiran replied:
For a moment I feared that you have discovered a security bug.

Thankfully the code is robust and doesn't allow you to use ``PROTOCOL_TLS_CLIENT`` with default arguments of ``wrap_socket`` and ``SSLSocket``. Both ``ssl.wrap_socket(ssl_version=ssl.PROTOCOL_TLS_CLIENT)`` and ``ssl.SSLSocket(ssl_version=ssl.PROTOCOL_TLS_CLIENT)`` will both fail with error ``Cannot set verify_mode to CERT_NONE when check_hostname is enabled.``.

So the insecure default doesn't work here. I consider it a feature. :) I'm going to deprecate ``ssl.wrap_socket`` and ``ssl.SSLSocket()`` default constructor because they have more fundamental issues.

vadmium replied:
Okay, now I realize that the PROTOCOL_TLS_CLIENT mode is newer than the old _wrap_socket_ function. It was not clear that _wrap_socket_ doesn’t support this new mode, but that is separate to what you are trying to do here.

tiran committed:
Drop remark about anonymous ciphers

The documentation no longer mentions anonymous ciphers. They are totally
insecure and irrelevant for virtually all users. The documentation was
also wrong. For anonymous ciphers, CERT_NONE, CERT_OPTIONAL and
CERT_REQUIRED all behaved the same.

Internally, CERT_REQUIRED = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT
but the last flag is ignored in client mode. When a anonymous cipher is
used, clients ignore missing server certs.

Signed-off-by: Christian Heimes <christian@python.org>
